src/covid_model_seiir_pipeline/workflow_tools/specification.py

```diff
@@ -0,0 +1,148 @@
+"""Primitives for construction jobmon workflow specifications."""
+import abc
+import re
+from typing import Dict, Type, TypeVar, Union
+
+from loguru import logger
+
+
+DEFAULT_PROJECT = 'proj_covid'
+DEFAULT_QUEUE = 'd.q'
+
+_TaskSpecDict = Dict[str, Union[str, int]]
+
+
+class TaskSpecification(abc.ABC):
+    """Validation wrapper for specifying task execution parameters.
+
+    This class is intended to outline the parameter space for task
+    specification and provide sanity checks on the passed parameters so
+    that we fail fast and informatively.
+
+    Subclasses are meant to inherit and provide default values for the
+    class variables. At runtime those defaults will be used to fill values
+    not provided in the specification file when the specification file
+    is parsed and then the full task specification will be validated against
+    the values in the `validate` method of this class.
+
+    """
+    default_max_runtime_seconds: int
+    default_m_mem_free: str
+    default_num_cores: int
```

Comment:
Why is this an ABC?

Recommend `__init_subclass__` for ensuring these defaults are filled in by subclasses.

Reply:
It's an abc because it has no defaults and therefore shouldn't be used directly. Your solution is better though.  I always forget about that hook.